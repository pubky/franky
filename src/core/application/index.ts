export * from './bootstrap';
export * from './tag';
export * from './post';
export * from './profile';
export * from './user';
export * from './stream';
export * from './notification';
