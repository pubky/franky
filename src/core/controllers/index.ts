export * from './user';
export * from './auth';
export * from './stream';
// export * from './post';
export * from './post';
export * from './postMock';
export * from './user';
