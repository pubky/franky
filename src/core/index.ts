export * from './controllers';
export * from './database';
export * from './models';
export * from './services';
export * from './pipes';
export * from './stores';
