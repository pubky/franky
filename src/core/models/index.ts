export * from './post';
export * from './user';
export * from './stream';
export * from './shared';
export * from './notification';
export * from './models.types';
export * from './models.defaults';
