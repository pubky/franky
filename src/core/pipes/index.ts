// Pipes have two typical use cases:
// 1. validation: evaluate input data and if valid, simply pass it through unchanged; otherwise, throw an exception
// 2. Normalization: transform input data to the desired PubkyAppSpecs format

export * from './pipes.builder';
export * from './pipes.types';
export * from './file';
export * from './post';
export * from './tag';
export * from './user';
export * from './follow';
export * from './mute';
export * from './notification';
// Re-export pubky-app-specs types
export { PubkyAppPostKind } from 'pubky-app-specs';
