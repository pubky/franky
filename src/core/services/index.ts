export * from './nexus';
export * from './homeserver';
export * from './localdb';
