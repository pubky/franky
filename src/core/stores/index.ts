export * from './onboarding';
export * from './profile';
