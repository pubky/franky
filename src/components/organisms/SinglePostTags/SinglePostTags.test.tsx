import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import { SinglePostTags } from './SinglePostTags';
import { useLiveQuery } from 'dexie-react-hooks';
import * as Core from '@/core';

// Mock @/libs - use actual implementations
vi.mock('@/libs', async (importOriginal) => {
  const actual = await importOriginal<typeof import('@/libs')>();
  return {
    ...actual,
  };
});

// Mock dexie-react-hooks
vi.mock('dexie-react-hooks', () => ({
  useLiveQuery: vi.fn(),
}));

// Mock the atoms
vi.mock('@/atoms', () => ({
  Container: vi.fn(({ children, className }) => (
    <div data-testid="container" className={className}>
      {children}
    </div>
  )),
  Input: vi.fn(({ placeholder, className, value, onChange, onKeyDown, disabled }) => (
    <input
      data-testid="input"
      placeholder={placeholder}
      className={className}
      value={value}
      onChange={onChange}
      onKeyDown={onKeyDown}
      disabled={disabled}
    />
  )),
  Badge: vi.fn(({ children, variant, className, onClick }) => (
    <div data-testid="badge" data-variant={variant} className={className} onClick={onClick}>
      {children}
    </div>
  )),
  Typography: vi.fn(({ children, size, className }) => (
    <span data-testid="typography" data-size={size} className={className}>
      {children}
    </span>
  )),
  Button: vi.fn(({ children, variant, size, className, onClick }) => (
    <button data-testid="button" data-variant={variant} data-size={size} className={className} onClick={onClick}>
      {children}
    </button>
  )),
  Link: vi.fn(({ children, href, className, style }) => (
    <a data-testid="link" href={href} className={className} style={style}>
      {children}
    </a>
  )),
  Avatar: vi.fn(({ children, className }) => (
    <div data-testid="avatar" className={className}>
      {children}
    </div>
  )),
  AvatarImage: vi.fn(({ src }) => <img data-testid="avatar-image" src={src} alt="" />),
  AvatarFallback: vi.fn(({ children }) => <div data-testid="avatar-fallback">{children}</div>),
}));

// Mock the core
vi.mock('@/core', async () => {
  const actual = await vi.importActual<typeof import('@/core')>('@/core');
  return {
    ...actual,
    useAuthStore: vi.fn(),
    db: {
      post_tags: {
        get: vi.fn().mockResolvedValue({ tags: [] }),
      },
    },
    TagController: {
      create: vi.fn(),
      delete: vi.fn(),
    },
  };
});

const mockUseLiveQuery = vi.mocked(useLiveQuery);
const mockUseAuthStore = vi.mocked(Core.useAuthStore);
const mockDbGet = vi.mocked(Core.db.post_tags.get);
const mockTagControllerCreate = vi.mocked(Core.TagController.create);
const mockTagControllerDelete = vi.mocked(Core.TagController.delete);

// Mock console.log and console.error
const mockConsoleLog = vi.spyOn(console, 'log').mockImplementation(() => {});
const mockConsoleError = vi.spyOn(console, 'error').mockImplementation(() => {});

describe('SinglePostTags', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    mockUseAuthStore.mockReturnValue('test-user-id');
    mockDbGet.mockResolvedValue({ tags: [] });
    mockTagControllerCreate.mockResolvedValue(undefined);
    mockTagControllerDelete.mockResolvedValue(undefined);
    mockUseLiveQuery.mockReturnValue([]);
  });

  it('renders with required postId prop', () => {
    render(<SinglePostTags postId="test-post-123" />);

    const containers = screen.getAllByTestId('container');
    expect(containers).toHaveLength(2);
  });

  it('handles input value changes', () => {
    render(<SinglePostTags postId="test-post-123" />);

    const input = screen.getByTestId('input');
    fireEvent.change(input, { target: { value: 'new-tag' } });

    expect(input).toHaveValue('new-tag');
  });

  it('handles Enter key to add tag', async () => {
    render(<SinglePostTags postId="test-post-123" />);

    const input = screen.getByTestId('input');
    fireEvent.change(input, { target: { value: 'new-tag' } });
    fireEvent.keyDown(input, { key: 'Enter' });

    await waitFor(() => {
      expect(mockTagControllerCreate).toHaveBeenCalledWith({
        taggedId: 'test-post-123',
        label: 'new-tag',
        taggerId: 'test-user-id',
        taggedKind: Core.TagKind.POST,
      });
    });
  });

  it('does not add tag with empty input', async () => {
    render(<SinglePostTags postId="test-post-123" />);

    const input = screen.getByTestId('input');
    fireEvent.change(input, { target: { value: '   ' } });
    fireEvent.keyDown(input, { key: 'Enter' });

    expect(mockTagControllerCreate).not.toHaveBeenCalled();
  });

  it('trims whitespace from input before adding tag', async () => {
    render(<SinglePostTags postId="test-post-123" />);

    const input = screen.getByTestId('input');
    fireEvent.change(input, { target: { value: '  new-tag  ' } });
    fireEvent.keyDown(input, { key: 'Enter' });

    await waitFor(() => {
      expect(mockTagControllerCreate).toHaveBeenCalledWith({
        taggedId: 'test-post-123',
        label: 'new-tag',
        taggerId: 'test-user-id',
        taggedKind: Core.TagKind.POST,
      });
    });
  });

  it('clears input after successful tag creation', async () => {
    render(<SinglePostTags postId="test-post-123" />);

    const input = screen.getByTestId('input');
    fireEvent.change(input, { target: { value: 'new-tag' } });
    fireEvent.keyDown(input, { key: 'Enter' });

    await waitFor(() => {
      expect(input).toHaveValue('');
    });
  });

  it('handles tag creation errors', async () => {
    const error = new Error('Tag creation failed');
    mockTagControllerCreate.mockRejectedValueOnce(error);

    render(<SinglePostTags postId="test-post-123" />);

    const input = screen.getByTestId('input');
    fireEvent.change(input, { target: { value: 'new-tag' } });
    fireEvent.keyDown(input, { key: 'Enter' });

    await waitFor(() => {
      expect(mockConsoleError).toHaveBeenCalledWith('Failed to add tag:', error);
    });
  });

  it('renders tags from data', () => {
    const mockTags = [
      {
        label: 'react',
        taggers_count: 3,
        relationship: true,
        taggers: ['user1', 'user2', 'user3'],
      },
      {
        label: 'javascript',
        taggers_count: 1,
        relationship: false,
        taggers: ['user4'],
      },
    ];
    mockUseLiveQuery.mockReturnValue(mockTags);

    render(<SinglePostTags postId="test-post-123" />);

    const badges = screen.getAllByTestId('badge');
    expect(badges).toHaveLength(3); // 2 tags + 1 mobile add button
  });

  it('handles tag toggle', async () => {
    const mockTags = [
      {
        label: 'react',
        taggers_count: 3,
        relationship: true,
        taggers: ['user1', 'user2', 'user3'],
      },
    ];
    mockUseLiveQuery.mockReturnValue(mockTags);

    render(<SinglePostTags postId="test-post-123" />);

    const badges = screen.getAllByTestId('badge');
    fireEvent.click(badges[0]); // Click on react tag

    await waitFor(() => {
      expect(mockTagControllerDelete).toHaveBeenCalledWith({
        taggedId: 'test-post-123',
        label: 'react',
        taggerId: 'test-user-id',
        taggedKind: Core.TagKind.POST,
      });
    });
  });

  it('handles tag toggle errors', async () => {
    const error = new Error('Tag toggle failed');
    mockTagControllerCreate.mockRejectedValueOnce(error);

    const mockTags = [
      {
        label: 'javascript',
        taggers_count: 1,
        relationship: false,
        taggers: ['user4'],
      },
    ];
    mockUseLiveQuery.mockReturnValue(mockTags);

    render(<SinglePostTags postId="test-post-123" />);

    const badges = screen.getAllByTestId('badge');
    fireEvent.click(badges[0]);

    await waitFor(() => {
      expect(mockConsoleError).toHaveBeenCalledWith('Failed to toggle tagger:', error);
    });
  });

  it('handles search button click', () => {
    const mockTags = [
      {
        label: 'react',
        taggers_count: 3,
        relationship: true,
        taggers: ['user1', 'user2', 'user3'],
      },
    ];
    mockUseLiveQuery.mockReturnValue(mockTags);

    render(<SinglePostTags postId="test-post-123" />);

    const buttons = screen.getAllByTestId('button');
    fireEvent.click(buttons[0]);

    expect(mockConsoleLog).toHaveBeenCalledWith('TODO: perform search with tag', 'react');
  });

  it('renders remaining count when more than MAX_VISIBLE_USERS', () => {
    const mockTags = [
      {
        label: 'react',
        taggers_count: 8,
        relationship: true,
        taggers: ['user1', 'user2', 'user3', 'user4', 'user5', 'user6', 'user7', 'user8'],
      },
    ];
    mockUseLiveQuery.mockReturnValue(mockTags);

    render(<SinglePostTags postId="test-post-123" />);

    expect(screen.getByText('+3')).toBeInTheDocument(); // 8 - 5 = 3 remaining
  });

  it('handles mobile add tag button click', () => {
    render(<SinglePostTags postId="test-post-123" />);

    const badges = screen.getAllByTestId('badge');
    const mobileAddButton = badges.find(
      (badge) => badge.className.includes('lg:hidden') && badge.className.includes('border-dashed'),
    );
    fireEvent.click(mobileAddButton!);

    expect(mockConsoleLog).toHaveBeenCalledWith('TODO: show add tag modal');
  });

  it('does not add tag when currentUserId is null', async () => {
    mockUseAuthStore.mockReturnValue(null);

    render(<SinglePostTags postId="test-post-123" />);

    const input = screen.getByTestId('input');
    fireEvent.change(input, { target: { value: 'new-tag' } });
    fireEvent.keyDown(input, { key: 'Enter' });

    expect(mockTagControllerCreate).not.toHaveBeenCalled();
  });

  it('does not toggle tag when currentUserId is null', async () => {
    mockUseAuthStore.mockReturnValue(null);

    const mockTags = [
      {
        label: 'javascript',
        taggers_count: 1,
        relationship: false,
        taggers: ['user4'],
      },
    ];
    mockUseLiveQuery.mockReturnValue(mockTags);

    render(<SinglePostTags postId="test-post-123" />);

    const badges = screen.getAllByTestId('badge');
    fireEvent.click(badges[0]);

    expect(mockTagControllerCreate).not.toHaveBeenCalled();
  });

  it('calls Core.db.post_tags.get with correct postId', async () => {
    render(<SinglePostTags postId="test-post-123" />);

    const callback = mockUseLiveQuery.mock.calls[0][0];
    await callback();

    expect(mockDbGet).toHaveBeenCalledWith('test-post-123');
  });
});

describe('SinglePostTags - Snapshots', () => {
  it('matches snapshot with no tags', () => {
    mockUseLiveQuery.mockReturnValue([]);
    const { container } = render(<SinglePostTags postId="test-post-123" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with tags', () => {
    const mockTags = [
      {
        label: 'react',
        taggers_count: 3,
        relationship: true,
        taggers: ['user1', 'user2', 'user3'],
      },
      {
        label: 'javascript',
        taggers_count: 1,
        relationship: false,
        taggers: ['user4'],
      },
    ];
    mockUseLiveQuery.mockReturnValue(mockTags);
    const { container } = render(<SinglePostTags postId="test-post-123" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with many taggers', () => {
    const mockTags = [
      {
        label: 'react',
        taggers_count: 8,
        relationship: true,
        taggers: ['user1', 'user2', 'user3', 'user4', 'user5', 'user6', 'user7', 'user8'],
      },
    ];
    mockUseLiveQuery.mockReturnValue(mockTags);
    const { container } = render(<SinglePostTags postId="test-post-123" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with different postId', () => {
    mockUseLiveQuery.mockReturnValue([]);
    const { container } = render(<SinglePostTags postId="different-post-id" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
