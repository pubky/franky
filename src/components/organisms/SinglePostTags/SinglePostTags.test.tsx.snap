// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SinglePostTags - Snapshots > matches snapshot with different postId 1`] = `
<div
  class="flex flex-col"
  data-testid="container"
>
  <div
    class="hidden lg:flex gap-2"
    data-testid="container"
  >
    <input
      class="flex-1"
      data-testid="input"
      placeholder="add tag"
      value=""
    />
  </div>
  <div
    class="flex flex-row lg:flex-col flex-wrap gap-2 lg:pt-2"
  >
    <div
      class="lg:hidden flex items-center gap-1 cursor-pointer hover:opacity-80 transition-opacity border-dashed h-6"
      data-testid="badge"
      data-variant="outline"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-plus h-3 w-3"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
        <path
          d="M12 5v14"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`SinglePostTags - Snapshots > matches snapshot with many taggers 1`] = `
<div
  class="flex flex-col"
  data-testid="container"
>
  <div
    class="hidden lg:flex gap-2"
    data-testid="container"
  >
    <input
      class="flex-1"
      data-testid="input"
      placeholder="add tag"
      value=""
    />
  </div>
  <div
    class="flex flex-row lg:flex-col flex-wrap gap-2 lg:pt-2"
  >
    <div
      class="flex flex-row gap-2 items-center"
    >
      <div
        class="flex items-center gap-1 cursor-pointer hover:opacity-80 transition-opacity"
        data-testid="badge"
        data-variant="default"
      >
        react
        <span
          class="opacity-70"
          data-size="sm"
          data-testid="typography"
        >
          8
        </span>
      </div>
      <button
        class="hidden lg:flex rounded-full w-8 h-8 p-0"
        data-size="sm"
        data-testid="button"
        data-variant="secondary"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-search h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m21 21-4.34-4.34"
          />
          <circle
            cx="11"
            cy="11"
            r="8"
          />
        </svg>
      </button>
      <div
        class="hidden lg:flex items-center -space-x-2"
      >
        <a
          class="relative"
          data-testid="link"
          href="/profile/user1"
          style="z-index: 10; filter: drop-shadow(0 2px 4px rgba(0, 0, 0, 0.3));"
        >
          <div
            class="w-8 h-8 cursor-pointer hover:opacity-80 transition-opacity"
            data-testid="avatar"
          >
            <img
              alt=""
              data-testid="avatar-image"
            />
            <div
              data-testid="avatar-fallback"
            >
              US
            </div>
          </div>
        </a>
        <a
          class="relative"
          data-testid="link"
          href="/profile/user2"
          style="z-index: 11; filter: drop-shadow(0 2px 4px rgba(0, 0, 0, 0.3));"
        >
          <div
            class="w-8 h-8 cursor-pointer hover:opacity-80 transition-opacity"
            data-testid="avatar"
          >
            <img
              alt=""
              data-testid="avatar-image"
            />
            <div
              data-testid="avatar-fallback"
            >
              US
            </div>
          </div>
        </a>
        <a
          class="relative"
          data-testid="link"
          href="/profile/user3"
          style="z-index: 12; filter: drop-shadow(0 2px 4px rgba(0, 0, 0, 0.3));"
        >
          <div
            class="w-8 h-8 cursor-pointer hover:opacity-80 transition-opacity"
            data-testid="avatar"
          >
            <img
              alt=""
              data-testid="avatar-image"
            />
            <div
              data-testid="avatar-fallback"
            >
              US
            </div>
          </div>
        </a>
        <a
          class="relative"
          data-testid="link"
          href="/profile/user4"
          style="z-index: 13; filter: drop-shadow(0 2px 4px rgba(0, 0, 0, 0.3));"
        >
          <div
            class="w-8 h-8 cursor-pointer hover:opacity-80 transition-opacity"
            data-testid="avatar"
          >
            <img
              alt=""
              data-testid="avatar-image"
            />
            <div
              data-testid="avatar-fallback"
            >
              US
            </div>
          </div>
        </a>
        <a
          class="relative"
          data-testid="link"
          href="/profile/user5"
          style="z-index: 14; filter: drop-shadow(0 2px 4px rgba(0, 0, 0, 0.3));"
        >
          <div
            class="w-8 h-8 cursor-pointer hover:opacity-80 transition-opacity"
            data-testid="avatar"
          >
            <img
              alt=""
              data-testid="avatar-image"
            />
            <div
              data-testid="avatar-fallback"
            >
              US
            </div>
          </div>
        </a>
        <div
          class="w-8 h-8 rounded-full bg-black text-white flex items-center justify-center cursor-pointer hover:opacity-80 transition-opacity text-xs font-medium"
          style="z-index: 15;"
        >
          +
          3
        </div>
      </div>
    </div>
    <div
      class="lg:hidden flex items-center gap-1 cursor-pointer hover:opacity-80 transition-opacity border-dashed h-6"
      data-testid="badge"
      data-variant="outline"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-plus h-3 w-3"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
        <path
          d="M12 5v14"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`SinglePostTags - Snapshots > matches snapshot with no tags 1`] = `
<div
  class="flex flex-col"
  data-testid="container"
>
  <div
    class="hidden lg:flex gap-2"
    data-testid="container"
  >
    <input
      class="flex-1"
      data-testid="input"
      placeholder="add tag"
      value=""
    />
  </div>
  <div
    class="flex flex-row lg:flex-col flex-wrap gap-2 lg:pt-2"
  >
    <div
      class="lg:hidden flex items-center gap-1 cursor-pointer hover:opacity-80 transition-opacity border-dashed h-6"
      data-testid="badge"
      data-variant="outline"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-plus h-3 w-3"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
        <path
          d="M12 5v14"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`SinglePostTags - Snapshots > matches snapshot with tags 1`] = `
<div
  class="flex flex-col"
  data-testid="container"
>
  <div
    class="hidden lg:flex gap-2"
    data-testid="container"
  >
    <input
      class="flex-1"
      data-testid="input"
      placeholder="add tag"
      value=""
    />
  </div>
  <div
    class="flex flex-row lg:flex-col flex-wrap gap-2 lg:pt-2"
  >
    <div
      class="flex flex-row gap-2 items-center"
    >
      <div
        class="flex items-center gap-1 cursor-pointer hover:opacity-80 transition-opacity"
        data-testid="badge"
        data-variant="default"
      >
        react
        <span
          class="opacity-70"
          data-size="sm"
          data-testid="typography"
        >
          3
        </span>
      </div>
      <button
        class="hidden lg:flex rounded-full w-8 h-8 p-0"
        data-size="sm"
        data-testid="button"
        data-variant="secondary"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-search h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m21 21-4.34-4.34"
          />
          <circle
            cx="11"
            cy="11"
            r="8"
          />
        </svg>
      </button>
      <div
        class="hidden lg:flex items-center -space-x-2"
      >
        <a
          class="relative"
          data-testid="link"
          href="/profile/user1"
          style="z-index: 10; filter: drop-shadow(0 2px 4px rgba(0, 0, 0, 0.3));"
        >
          <div
            class="w-8 h-8 cursor-pointer hover:opacity-80 transition-opacity"
            data-testid="avatar"
          >
            <img
              alt=""
              data-testid="avatar-image"
            />
            <div
              data-testid="avatar-fallback"
            >
              US
            </div>
          </div>
        </a>
        <a
          class="relative"
          data-testid="link"
          href="/profile/user2"
          style="z-index: 11; filter: drop-shadow(0 2px 4px rgba(0, 0, 0, 0.3));"
        >
          <div
            class="w-8 h-8 cursor-pointer hover:opacity-80 transition-opacity"
            data-testid="avatar"
          >
            <img
              alt=""
              data-testid="avatar-image"
            />
            <div
              data-testid="avatar-fallback"
            >
              US
            </div>
          </div>
        </a>
        <a
          class="relative"
          data-testid="link"
          href="/profile/user3"
          style="z-index: 12; filter: drop-shadow(0 2px 4px rgba(0, 0, 0, 0.3));"
        >
          <div
            class="w-8 h-8 cursor-pointer hover:opacity-80 transition-opacity"
            data-testid="avatar"
          >
            <img
              alt=""
              data-testid="avatar-image"
            />
            <div
              data-testid="avatar-fallback"
            >
              US
            </div>
          </div>
        </a>
      </div>
    </div>
    <div
      class="flex flex-row gap-2 items-center"
    >
      <div
        class="flex items-center gap-1 cursor-pointer hover:opacity-80 transition-opacity"
        data-testid="badge"
        data-variant="secondary"
      >
        javascript
        <span
          class="opacity-70"
          data-size="sm"
          data-testid="typography"
        >
          1
        </span>
      </div>
      <button
        class="hidden lg:flex rounded-full w-8 h-8 p-0"
        data-size="sm"
        data-testid="button"
        data-variant="secondary"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-search h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m21 21-4.34-4.34"
          />
          <circle
            cx="11"
            cy="11"
            r="8"
          />
        </svg>
      </button>
      <div
        class="hidden lg:flex items-center -space-x-2"
      >
        <a
          class="relative"
          data-testid="link"
          href="/profile/user4"
          style="z-index: 10; filter: drop-shadow(0 2px 4px rgba(0, 0, 0, 0.3));"
        >
          <div
            class="w-8 h-8 cursor-pointer hover:opacity-80 transition-opacity"
            data-testid="avatar"
          >
            <img
              alt=""
              data-testid="avatar-image"
            />
            <div
              data-testid="avatar-fallback"
            >
              US
            </div>
          </div>
        </a>
      </div>
    </div>
    <div
      class="lg:hidden flex items-center gap-1 cursor-pointer hover:opacity-80 transition-opacity border-dashed h-6"
      data-testid="badge"
      data-variant="outline"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-plus h-3 w-3"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
        <path
          d="M12 5v14"
        />
      </svg>
    </div>
  </div>
</div>
`;
